### ISSUE TITLE
Bizarre error message when the bound of a typevar is a union

### ISSUE THREAD

**Issue by kirbyfan64** [python/mypy#2314]
``` python
from typing import *

class X: pass

T = TypeVar('T', bound=Union[X, None])

class Y(Generic[T]):
    def __init__(self) -> None: pass
    attr: T
```

```
x.py: note: In function "__init__":
x.py: error: Type argument "T`1" of "Y" must be a subtype of "Union[x.X, builtins.None]"
```

If this is a user error, the error message should at least be a bit more like a, well, _normal_ error message.

**Comment by kirbyfan64** [comment]
Also, this is only an error with `strict-optional` enabled.

**Comment by gvanrossum** [comment]
Can you clarify what's bad about the error? Just the use of ``"T`1"`` instead of just `T`? Or the fact that there's nothing in `__init__` even referencing `T`? Or what?

**Comment by kirbyfan64** [comment]
@gvanrossum The whole thing. There's no line number, the ``` `1``` looks totally random, and the fact that `__init__` never references `T`.

Is this code supposed to even be allowed? I mean, it looks innocent enough...

**Comment by gvanrossum** [comment]
What does this say when using '--strict-optional'?

--Guido (mobile)

**Comment by kirbyfan64** [comment]
@gvanrossum The error message I posted. Without `strict-optional`, there's no error.

**Comment by rwbarton** [comment]
It's not really related to `strict-optional`. You get the same behavior without it by changing the union to `Union[X, W]` for another class `W`. The program is correct and mypy should accept it.

Mypy thinks the only ways for a type `t` to be a subtype of a union is if `t` is either itself a union or `t` is a subtype of one of the items in the union. But here we have another way, `t` is a type variable `T` whose upper bound is a union. This probably affects other uses of type variables bounded by unions (not necessarily having to do with generic classes).

The error arises while computing the type of `__init__`: the `self` argument has type `Y[T]` and therefore mypy wants to check that the argument `T` is compatible with the bound on the parameter of `Y`. Of course it should be, but mypy thinks it isn't because of the above issue.

This should be easy to fix, I'll submit a PR.

**Comment by rwbarton** [comment]
Hmm, it's actually not quite as easy as I had hoped. The problem is there are two ways for a type variable to be a subtype of a union:
```
    T <: Union[X, Y]
```
if T is a type variable with an upper bound `Union[X, Y]`, but also
```
    T <: Union[T, S]
```
for any type variable `T`, and if we replace `T` by its upper bound on the left hand side, we will no longer be able to prove this. I guess we just have to try both possibilities.

**Comment by ilevkivskyi** [comment]
This is fixed on master, it looks like it was fixed by #3039
